# Command used in Chapter 5 for stand-alone bioinformatic tools: H. influenzae from Indonesia in carriage and disease

## 1) FastQC: https://github.com/s-andrews/FastQC
## Program was run through https://usegalaxy.eu/

### --- ###


## 2) SPAdes: https://github.com/ablab/spades
## Program was run through the University of Oxford ARC computer cluster

module load SPAdes/3.15.2-GCC-10.2.0

CHDIR="/working/dir"
genome="Genome ID" # genome/isolate name
OUTDIR="/dir/for/output"

spades.py -k 31,43,55,77 --isolate --disable-gzip-output --pe1-1 ${CHDIR}/${genome}_1.fastq \
	--pe1-2 ${CHDIR}/${genome}_2.fastq -o ${OUTDIR}/${genome} --threads 48

### --- ###


## 3) QUAST: https://github.com/ablab/quast
## Program was run through the University of Oxford ARC computer cluster

module load QUAST/5.0.2-foss-2020a-Python-3.8.2

INDIR="/dir/to/fasta/files"
OUTDIR="/dir/for/quast_output"

for f in $(cat ./fasta.txt); do # fasta.txt contains list of fasta file names
	quast.py ${INDIR}/$f -o ${OUTDIR}/$f;
done

### --- ###


## 4) RAxML: https://github.com/stamatak/standard-RAxML, and
## 5) ClonalFrameML: https://github.com/xavierdidelot/clonalframeml/wiki
## Program was run through the University of Oxford ARC computer cluster

module load RAxML/8.2.12-gompi-2020a-hybrid-avx2
module load ClonalFrameML/1.12-foss-2022a

infile="/dir/to/core_alignment.fasta" # core alignment from cgMLST scheme, generated by PubMLST Genome comparator

raxmlHPC -m GTRGAMMA -p 184029 -T 48 -N 5 -s ${infile} -n hinf_dataset_core.nwk
ClonalFrameML /dir/to/RAxML_output/RAxML_bestTree.hinf_dataset_core.nwk ${infile} hinf_dataset_CFML

### --- ###


## 6) VFDB BLAST: https://www.mgc.ac.cn/VFs/download.htm
## BLAST search was done through PubMLST Genome comparator

### --- ###


## 7) chewBBACA: https://chewbbaca.readthedocs.io/en/latest/
## Program was run locally

chewBBACA.py PrepExternalSchema -g /path/to/ExternalSchemaFolder -o /dir/chewbbaca_schema # ExternalSchemaFolder contains fasta files of VF nucleotide sequence (1 VF-gene per file)
chewBBACA.py AlleleCall -i /dir/fasta -g /dir/chewbbaca_schema -o ./chewbbaca_allele_call --cpu 6 --output-novel >> output.log

### --- ###


## 8) BLAST search
## Program was run through the University of Oxford ARC computer cluster

module load BLAST+/2.14.0-gompi-2022a

makeblastdb -in Hinf_plasmid_sequence.fasta -parse_seqids -title "Hinf_plasmid_db" -dbtype nucl -out Hinf_plasmid_db
	## Hinf_plasmid_sequence.fasta is a fasta file containing H. influenzae plasmid sequences retrieved from NCBI nucleotide database

for f in $(cat $fasta/fastafiles.txt); do # text file contains list of isolates' fasta file names
    blastn -query $fasta/${f} -db Hinf_plasmid_db -perc_identity 90 -outfmt 11 -num_threads 8 -out ${f%%.*}_hinf_plasmid.asn;
    blast_formatter -archive ${f%%.*}_hinf_plasmid.asn -outfmt 5 > ${f%%.*}_hinf_plasmid.xml;
    blast_formatter -archive ${f%%.*}_hinf_plasmid.asn -outfmt 6 > ${f%%.*}_hinf_plasmid.tsv;
done